### PR TITLE
Use Firebase REST API in the `firebase-web`

### PR DESCRIPTION
**Pull request by dmitrykuzmin**
Previously, the Firebase Admin SDK was used in order to access the Firebase database in the `firebase-web` module.

This caused a number of build issues that could not be easily resolved.

Now, the `firebase-web` uses [Firebase REST API](https://firebase.google.com/docs/reference/rest/database/) to operate on the database. All read/write operations became synchronous and are performed via [Google HTTP Client](https://github.com/googleapis/google-http-java-client) calls.

The `spine-dev` database security is currently changed to "free for all", so its API is accessible from tests. This issue is to be addressed soon.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
@armiol can you please TAL. I will be able to resolve coverage issues tomorrow but it's unlikely to be significant changes.

**Comment by dmitrykuzmin** [comment]
@armiol PTAL.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseClient.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     *
+     * <p>If the node exists, the new value will be added to the existing entries (i.e. the node
+     * content is <strong>not</strong> overwritten).
+     *
+     * @param nodePath
+     *         the path to the node in the Firebase database
+     * @param value
+     *         the value to add to the node
+     */
+    void addValue(FirebaseDatabasePath nodePath, FirebaseNodeValue value);
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseClient.java]: Can we please name it `append`?

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * Creates a {@linkplain io.spine.web.firebase.FirebaseRestClient firebase client} which
+     * operates via the Firebase REST API.
+     *
+     * @param databaseUrl
+     *         the URL of the database on which the client operates
+     * @return the new instance of {@code FirebaseRestClient}
+     */
+    public static FirebaseClient restClient(String databaseUrl) {
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseClientFactory.java]: I think we should validate the input value first. 

Ideally, it shouldn't be a `String`, rather a tiny type — then the validation goes to the tiny type factory method.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+     * <p>The different environment requires different {@linkplain
+     * com.google.api.client.http.HttpTransport HTTP transport} to operate.
+     *
+     * <p>For more information on this, see
+     * <a href="https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.20.0/com/google/api/client/http/HttpTransport">
+     * HttpTransport docs</a>.
+     */
+    private static FirebaseClient forCurrentEnv(String databaseUrl) {
+        try {
+            Class.forName(SystemProperty.class.getName());
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseClientFactory.java]: If I guessed the purpose correctly... We have `ServerEnvironment.isAppEngine()` that looks better for the environment detection.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseNodeValue.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+     *
+     * <p>The key for the child is generated via
+     * {@linkplain com.google.firebase.database.utilities.PushIdGenerator standard Firebase
+     * generation mechanism}.
+     *
+     * @return the generated key under which the data was stored
+     * @see ChildKeyGenerator
+     */
+    @CanIgnoreReturnValue
+    String pushData(String data) {
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseNodeValue.java]: I think it's more like `addChild`, similar to below. Then you'll have two overloads with almost exactly the same semantics.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseRestClient.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+     * Updates the database node with the given value.
+     *
+     * <p>Common entries are overwritten.
+     */
+    private void update(GenericUrl nodeUrl, HttpContent value) {
+        requestExecutor.patch(nodeUrl, value);
+    }
+
+    private GenericUrl toNodeUrl(FirebaseDatabasePath nodePath) {
+        String url = String.format(NODE_URL_FORMAT, databaseUrl, nodePath);
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseRestClient.java]: It seems to me that the performance could be better here. `databaseUrl` is the `final` field, and you can format it once the instance is created. So the `nodePath` is the only value to use in formatting.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseNodeValue.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        this.value = value;
+    }
+
+    /**
+     * Creates an empty {@code FirebaseNodeValue}.
+     *
+     * <p>This is not equivalent to the {@code null} value, an empty value is supposed to be
+     * filled with entries at some point after the creation.
+     */
+    FirebaseNodeValue() {
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply firebase-web/src/main/java/io/spine/web/firebase/FirebaseNodeValue.java]: Also, it is a good practice to have a single primary constructor and call it from all the other constructors, i.e. 
```java
private FirebaseNodeValue(JsonObject value) { // <- Primary ctor.
    this.value = value;
}

private FirebaseNodeValue() { // <- Secondary ctor.
    this(new JsonObject());
}
```

**Pull request merged**